cmd/thv-operator/controllers/mcpregistry_controller.go

```diff
@@ -211,16 +392,443 @@ func (r *MCPRegistryReconciler) finalizeMCPRegistry(ctx context.Context, registr
+		// Deployment doesn't exist, create it
+		desired, buildErr := r.buildRegistryAPIDeployment(mcpRegistry, sourceHandler)
+		if buildErr != nil {
+			return nil, fmt.Errorf("failed to build deployment configuration: %w", buildErr)
+		}
+
+		// Set owner reference
+		if ownerErr := controllerutil.SetControllerReference(mcpRegistry, desired, r.Scheme); ownerErr != nil {
+			return nil, fmt.Errorf("failed to set controller reference: %w", ownerErr)
+		}
+
+		if createErr := r.Create(ctx, desired); createErr != nil {
+			return nil, fmt.Errorf("failed to create deployment %s: %w", deploymentName, createErr)
+		}
+
+		ctxLogger.Info("Deployment created successfully", "deployment", deploymentName)
+		return desired, nil
+	} else if err != nil {
+		// Some other error occurred
+		return nil, fmt.Errorf("failed to get deployment %s: %w", deploymentName, err)
+	}
+
+	// TODO: Implement deployment updates when needed (e.g., when config hash changes)
+	// For now, just return the existing deployment to avoid endless reconciliation loops
+	ctxLogger.Info("Deployment already exists, skipping update", "deployment", deploymentName)
+	return deployment, nil
+}
+
+// buildRegistryAPIService creates and configures a Service object for the registry API.
+// This function handles all service configuration including labels, ports, and selector.
+// It returns a fully configured ClusterIP service ready for Kubernetes API operations.
+func buildRegistryAPIService(mcpRegistry *mcpv1alpha1.MCPRegistry) *corev1.Service {
+	// Generate service name using the established pattern
+	serviceName := fmt.Sprintf("%s-api", mcpRegistry.Name)
+
+	// Define labels using common function
+	labels := labelsForRegistryAPI(mcpRegistry, serviceName)
+
+	// Define selector to match deployment pod labels
+	selector := map[string]string{
+		"app.kubernetes.io/name":      serviceName,
+		"app.kubernetes.io/component": "registry-api",
+	}
+
+	// Create service specification
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceName,
+			Namespace: mcpRegistry.Namespace,
+			Labels:    labels,
+		},
+		Spec: corev1.ServiceSpec{
+			Type:     corev1.ServiceTypeClusterIP,
+			Selector: selector,
+			Ports: []corev1.ServicePort{
+				{
+					Name:       "http",
+					Port:       8080,
+					TargetPort: intstr.FromInt32(8080),
+					Protocol:   corev1.ProtocolTCP,
+				},
+			},
+		},
+	}
+
+	return service
+}
+
+// ensureService creates or updates the registry-api Service for the MCPRegistry.
+// This function handles the Kubernetes API operations (Get, Create, Update) and delegates
+// service configuration to buildRegistryAPIService.
+func (r *MCPRegistryReconciler) ensureService(
+	ctx context.Context,
+	mcpRegistry *mcpv1alpha1.MCPRegistry,
+) (*corev1.Service, error) {
+	ctxLogger := log.FromContext(ctx).WithValues("mcpregistry", mcpRegistry.Name)
+
+	// Build the desired service configuration
+	service := buildRegistryAPIService(mcpRegistry)
```

Comment:
maybe I can't follow the flows in the review, but it seems that the `ensureDeployment` method first checks for the existence of the deployment before invoking `buildRegistryAPIDeployment`, while here we do the other way: first we invoke `buildRegistryAPIService` then check for the existence of the service.

Reply:
you're right I changed them to be symmetric